examples/mem-correctness-test/src/main.rs

```diff
@@ -0,0 +1,267 @@
+/* This test based on the following pseudo code and tests correctness of our
+ * new allocator snmalloc
+
+   for 0..num_threads {
+    loop {
+        let mut regions: Vec<(Box<[u8], u8>)>;
+        let mut mem_used = 0u64;
+        match rnd % 4 {
+           0 => // Check area
+           1..2 => // Alloc random area if less than x% of the heap is used (not 100% to
+                   // account for fragmentation, ...) and check area.
+           3 => // Free random area
+        }
+    }
+}
+ * So this basically runs for a long time and should never crash
+ */
+
+
+use core::arch::asm;
+use num_cpus;
+use rand::Rng;
+use std::mem;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Condvar, Mutex};
+use std::thread;
+use std::time::Instant;
+
+extern "C" {
+    static HEAP_BASE: u64;
+    static HEAP_SIZE: usize;
+}
+
+const PAGE_SIZE: usize = 4096;
+const TO_KB: usize = 1024;
+const TO_MB: usize = TO_KB * 1024;
+const TO_GB: usize = TO_MB * 1024;
+
+const NUM_OPERATION_CHOICES: usize = 4;
```

Comment:
why don't you create an array of tester functions: fn traverse_and_check_buffer(buf: &Vec<u32>)
and in the worker thread we just go by the number of elements in the array. 
this way we can:
- expand the test variety by adding more operations
- change the test type weight/ratio by adding more functions of the same type
- remove test types, for debug purposes or when they become obsolete, or for whatever other reason
- other fun stuff

Reply:
I like this idea (as a future enhancement), but I would add that the usual way to have different weights for different functions is not to duplicate their entries in the table but rather to have an explicit weight associated with each table entry, and then randomly choose an activity taking the weights into account.